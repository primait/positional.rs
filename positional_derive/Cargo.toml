[package]
name = "positional_derive"
version = "0.1.1"
edition = "2021"
description = "macros for positional crate"
license = "MIT"

[lib]
proc-macro = true

[dependencies]
syn = { version = "1.0.95", features = ["extra-traits", "full"] }
quote = "1.0"
proc-macro2 = "1.0.39"
proc-macro-error = "1.0.4"
trybuild = "1.0.61"

