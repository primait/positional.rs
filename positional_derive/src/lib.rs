use proc_macro::TokenStream;
use proc_macro_error::proc_macro_error;

mod analyze;
mod codegen;
mod lower;
mod parse;

use analyze::analyze;
use codegen::codegen;
use lower::{lower, ImplBlockType};
use parse::{parse, Ast};

/// Add to structs to make them deserializable from positional rows
#[proc_macro_derive(FromPositionalRow, attributes(field))]
pub fn from_positional_row(tokens: TokenStream) -> TokenStream {
    let ast = parse(tokens.into());
    let model = analyze(ast);
    let ir = lower(model);
    let rust = codegen(ir, ImplBlockType::From);
    rust.into()
}

/// Add to structs to make them serializable into positional rows
#[proc_macro_derive(ToPositionalRow, attributes(field))]
#[proc_macro_error]
pub fn to_positional_row(tokens: TokenStream) -> TokenStream {
    let ast = parse(tokens.into());
    let model = analyze(ast);
    let ir = lower(model);
    let rust = codegen(ir, ImplBlockType::To);
    rust.into()
}
