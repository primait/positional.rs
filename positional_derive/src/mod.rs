mod analyze;
mod codegen;
mod lower;

pub use analyze::{analyze, FieldAlignment};
pub use codegen::{codegen, Rust};
pub use lower::{lower, Field, ImplBlockType};
