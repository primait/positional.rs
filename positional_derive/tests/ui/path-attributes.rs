use positional_derive::ToPositionalRow;

#[derive(ToPositionalRow)]
struct MyData {
    #[field]
    name: String,
}

fn main() {}
