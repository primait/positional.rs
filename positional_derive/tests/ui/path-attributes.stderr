error: failed to parse field attribute

         = note: expected attribute arguments in parentheses: #[field(...)]

 --> tests/ui/path-attributes.rs:5:5
  |
5 |     #[field]
  |     ^^^^^^^^
